services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala

```diff
@@ -1,29 +1,96 @@
 package cromwell.services.metadata.impl
 
-import akka.actor.{Actor, ActorLogging, Props}
+import akka.actor.{ActorLogging, LoggingFSM, Props}
 import cromwell.core.Dispatcher.ServiceDispatcher
 import cromwell.services.SingletonServicesStore
-import cromwell.services.metadata.MetadataService.{MetadataPutAcknowledgement, MetadataPutFailed, PutMetadataAction}
+import cromwell.services.metadata.MetadataEvent
+import cromwell.services.metadata.MetadataService.PutMetadataAction
+import cromwell.services.metadata.impl.WriteMetadataActor.{WriteMetadataActorData, WriteMetadataActorState}
 
+import scala.concurrent.duration._
 import scala.util.{Failure, Success}
 
-object WriteMetadataActor {
-  def props() = Props(new WriteMetadataActor()).withDispatcher(ServiceDispatcher)
+final case class WriteMetadataActor(batchRate: Int, flushRate: FiniteDuration)
+  extends LoggingFSM[WriteMetadataActorState, WriteMetadataActorData] with ActorLogging with
+  MetadataDatabaseAccess with SingletonServicesStore {
+  import WriteMetadataActor._
+
+  implicit val ec = context.dispatcher
+
+  override def preStart(): Unit = {
+    context.system.scheduler.schedule(0.seconds, flushRate, self, ScheduledFlushToDb)
+    super.preStart()
+  }
+
+  startWith(WaitingToWrite, NoEvents)
+
+  when(WaitingToWrite) {
+    case Event(PutMetadataAction(events), curData) =>
+      curData.addEvents(events) match {
+        case data@HasEvents(e) if e.size > batchRate => goto(WritingToDb) using data
+        case e => stay using e
+      }
+    case Event(ScheduledFlushToDb, curData) =>
+      log.debug("Initiating periodic metadata flush to DB")
+      goto(WritingToDb) using curData
+  }
+
+  when(WritingToDb) {
+    case Event(ScheduledFlushToDb, curData) => stay using curData
+    case Event(PutMetadataAction(events), curData) => stay using curData.addEvents(events)
+    case Event(FlushBatchToDb, curData) =>
+      curData match {
```

Comment:
Looking at this I don't particularly care for it either but I don't understand what you're after w/ that comment